feat: added azure blob instrumentation

athirasomanath opened the pull request:
This feature focuses on  instrumenting the classes blobClient and blockBlobClient in azure blob storage:
https://instana.kanbanize.com/ctrl_board/56/cards/146704/details/

Methods instrumented for blob:

- Upload
- Download
- Delete

NOTE: Instrumenting BlockBlobClient and BlobClient give easy parsing for proper data required for span creation such as blobname, container name and operation. If using sendOperation, we need to parse it from the url and it doesn't seem neat. Also sendOperation is from @azure/core-http , and even though it is used in almost every operations in storage package, it doesn't act as common point of instrumentation when it comes to a major package like cosmosDB

Things to cover

- Page Blob class
- Container Class for instrumenting container functionalities

kirrg001 reviewed (changes requested):
Hey! 

I need to do another review iteration with fresh eyes. Please take a look at my first comments 👍 

Here are some additional findings:
- ESM app is missing. Please add an ESM app.mjs and check if azure blob storage works with ESM - [see](https://github.com/instana/nodejs/blob/main/CONTRIBUTING.md#adding-tests).
- I could not find the ticket reference in the PR nor in the commits.

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,343 @@
+  const _blockBlobClient = containerClient.getBlockBlobClient(_blobName);
+  try {
+    const downloadBlobResponse = await _blockBlobClient.downloadToFile(localFilePath);
+    console.log(`Downloaded blob to ${localFilePath}:`);
```
Please reduce the number of console.* in all files 👍

athirasomanath replied:
removed unwanted logs

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,444 @@
+        }
+    });
+    ProcessControls.setUpHooks(controls);
+    if (!process.env.ACCOUNT_KEY) {
```
Initially you are assigning process.env.ACCOUNT_KEY to a variable, the here you are accessing process.env again.

athirasomanath replied:
used assigned variable

athirasomanath commented on packages/core/src/tracing/instrumentation/cloud/azure/blob.js:
```diff
@@ -0,0 +1,167 @@
+}
+function instrumentBlockBlob(blockBlobClient) {
+    shimmer.wrap(blockBlobClient.prototype, 'upload', shimUpload);
+    shimmer.wrap(blockBlobClient.prototype, 'stageBlock', shimUpload);
```
When data length is more than that of specified as maxSingleShotSize , stageBlock will be called to upload blocks.
it is tested in [testcase](https://github.com/instana/nodejs/blob/fe35e3d476117c49f24725e3728c7a56a9305895/packages/collector/test/tracing/cloud/azure/blob/app.js#L254)

kirrg001 replied:
Okay, but if I comment out:

// shimmer.wrap(blockBlobClient.prototype, 'stageBlock', shimUpload);

all tests still pass. So something is wrong?

athirasomanath replied:
Testsa re passing because, the span parameters for **stage block  upload** doesnt really differ from that of **upload**.  The value of op in span data is **upload** in both case, since the opertaion is ultimately blob upload for the user. Any change required in span data that will be useful to end user?

kirrg001 replied:
If the `stageBlock` function would have been called, the test should fail if I comment out this line?
I do not understand your reasoning.

athirasomanath replied:
My mistake. The testfile that got pushed to repo doesnt invoke the stageblock (file size is lesser in KB ). fixed that in testcase by adjusting maxSingleShotSize.

kirrg001 commented on packages/core/src/tracing/instrumentation/cloud/azure/blob.js:
```diff
@@ -0,0 +1,167 @@
+}
+function instrumentContainer(blobClient) {
+    shimmer.wrap(blobClient.prototype, 'delete', shimDelete);
+    shimmer.wrap(blobClient.prototype, 'download', shimDownload);
```
Please add a comment and describe how it works 👍

athirasomanath replied:
sure

athirasomanath replied:
added comments

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,343 @@
+        `Error uploading document ${blockBlobClient.name} to container ${blockBlobClient.containerName}`
+      );
+    } else {
+      console.log('uploaded successsfully');
```
used log function

athirasomanath commented on packages/core/src/tracing/instrumentation/cloud/azure/blob.js:
```diff
@@ -0,0 +1,167 @@
+            op
+        };
+        let originalCallback;
+        let callbackIndex = -1;
```
removed callback logic

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,444 @@
+
+const { createContainer, deleteContainer, miniNodeJsVer } = require('./util');
+const { BlobServiceClient } = require('@azure/storage-blob');
+const containerName = `test-${uuid()}`;
```
changed the generic name

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,444 @@
+            CONN_STR: connStr,
+            STORAGE_ACC: storageAccount,
+            ACC_KEY: accountKey,
+            CONT_CLIENT: containerClient
```
removed unwanted variable

athirasomanath commented on packages/core/src/tracing/instrumentation/cloud/azure/blob.js:
```diff
@@ -0,0 +1,167 @@
+            blob,
+            accountName,
+            op
+        };
```
fixed attribute names

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,343 @@
+
+const express = require('express');
+// const morgan = require('morgan');
+// const request = require('request-promise-native');
```
removed commented lines

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,444 @@
+const globalAgent = require('../../../../globalAgent');
+const { fail } = expect;
+
+const { createContainer, deleteContainer, miniNodeJsVer } = require('./util');
```
fixed typo

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,444 @@
+const { BlobServiceClient } = require('@azure/storage-blob');
+const containerName = `test-${uuid()}`;
+const storageAccount = process.env.STORAGE_ACCOUNT;
+const accountKey = process.env.ACCOUNT_KEY;
```
made changes

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,343 @@
+/*
+ * (c) Copyright IBM Corp. 2021
+ * (c) Copyright Instana Inc. and contributors 2018
```
fixed

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,343 @@
+/*
+ * (c) Copyright IBM Corp. 2021
```
resolved

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,444 @@
+                    path: '/upload-err'
+                });
+            const spans = await agentControls.getSpans();
+            await verifyOp({
```
The verify function is to check the attributes of all azure spans created while running a testcase. verifyOp tests specifically the operation  name and number of spans created for the operation intended in a testcase.
Eg: in a testcase for testing instrumentation of download, there will be spans created for uploading that blob as well. so in verify , all such spans are checked and  verifyOps confirms only one span with download op is created

athirasomanath commented on packages/core/src/tracing/instrumentation/cloud/azure/blob.js:
```diff
@@ -0,0 +1,167 @@
+    shimmer.wrap(blockBlobClient.prototype, 'stageBlock', shimUpload);
+}
+function shimUpload(original) {
+    return function () {
```
resolved

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,444 @@
+const containerName = `test-${uuid()}`;
+const storageAccount = process.env.STORAGE_ACCOUNT;
+const accountKey = process.env.ACCOUNT_KEY;
+const endPoint = 'core.windows.net';
```
added comment

kirrg001 replied:
Thanks. The comment does not really help to make it clear. Do you have a link which describes how it works?

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,444 @@
+        }
+    });
+    ProcessControls.setUpHooks(controls);
+    if (!process.env.ACCOUNT_KEY) {
```
resolved

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,444 @@
+let mochaSuiteFn = supportedVersion(process.versions.node) ? describe : describe.skip;
+mochaSuiteFn = semver.lt(process.versions.node, miniNodeJsVer) ? describe.skip : mochaSuiteFn;
+
+mochaSuiteFn('tracing/cloud/azure/blob', function () {
```
added comment

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,444 @@
+            span => expect(span.ec).to.equal(withError ? 1 : 0),
+            span => expect(span.data).to.exist,
+            span => expect(span.data[dataProperty || spanName]).to.be.an('object'),
+            span => expect(span.data[dataProperty || spanName].accountName).to.exist,
```
resolved

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,444 @@
+                    method: 'GET',
+                    path: '/upload'
+                });
+            const spans = await agentControls.getSpans();
```
added retry

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,325 @@
+const data1 = fs.readFileSync(filePath);
+const accountKey = process.env.ACC_KEY;
+const connStr = process.env.CONN_STR;
+const storageAccount = process.env.STORAGE_ACC;
```
[The variable names need to be changed as well.](https://github.com/instana/nodejs/pull/967#discussion_r1429915939)

athirasomanath replied:
resolved

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,498 @@
+const storageAccount = process.env.BLOB_STORAGE_ACCOUNT;
+const accountKey = process.env.BLOB_ACCOUNT_KEY;
+
+// endPoint is assigned with the default costant value for EndpointSuffix in connection string of an azure account
```
type `costant`

athirasomanath replied:
fixed typo

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,477 @@
+      dirname: __dirname,
+      useGlobalAgent: true,
+      env: {
+        CONTAINER_NAME: containerName,
```
Could you please add a prefix such as `AZURE_CONTAINER_NAME`?
See [example](https://github.com/instana/nodejs/blob/v3.0.0/packages/collector/test/tracing/database/couchbase/test.js#L106).

athirasomanath replied:
fixed

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,328 @@
+
+const blobServiceClient = BlobServiceClient.fromConnectionString(connStr);
+const containerName = process.env.AZURE_CONTAINER_NAME;
+const blobName = 'first.pdf';
```
As far as I understand, on our storage `teamnodejstracer` there is already a file called `first.pdf`.
If we need to use a new storage, the test would no longer work. Is that correct?

athirasomanath replied:
Every time the test suite runs, it will create a new container in the storage account and do file operations  specified in tests  in that container and in the end the container is deleted. In testcases we are doing the file upload whenever it requires (not storing any file in prior). The existing file is part of some test runs that might have stopped in the middle. I have deleted that and performed rerun to make sure that testcases still work.

kirrg001 replied:
Thanks 🙏

athirasomanath replied:
> As far as I understand, on our storage `teamnodejstracer` there is already a file called `first.pdf`. If we need to use a new storage, the test would no longer work. Is that correct?

We are not keeping any files in prior. The existing file is part of some cancelled test run. I have removed it as tested. For every test run , we will create a container, perform file operations and will delete container at the end of test suite. All file upload operations is done in individual testcases whenever it is required. And for a new storage account, we just need to configure the account name and key in the environment variables

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,477 @@
+      }
+    });
+    ProcessControls.setUpHooks(controls);
+    this.timeout(config.getTestTimeout());
```
The configuration of a custom timeout always needs to be on the first line.
The `ProcessControls` are registering before and after hooks as well. If we do not increase the timeout in the first line, they can easily run into the default timeout (which is two seconds).
This timeout can happen when running code coverage tools, which wrap our tests.

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,477 @@
+        path: '/uploadDataBlock'
+      });
+      await testUtils.retry(async () => {
+        const spans = await agentControls.getSpans();
```
Please check the number of spans, see [my comment.](https://github.com/instana/nodejs/pull/967#discussion_r1430014856)

kirrg001 replied:
> opSpanCount: 1

Please add the number of *all* spans. 
e.g. 2 http calls + azure call + fs call = 4

We always expect one azure span, please use "expectExactlyOneMatching".

athirasomanath replied:
resolved

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,477 @@
+        span => expect(span.data[dataProperty || spanName].op).to.exist
+      ]);
+    }
+    async function verifyOp({ dataProperty, op, n, spans }) {
```
I still don't understand why we need two verify functions. You could check the operation in the verify function.

athirasomanath replied:
resolved

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,328 @@
+    await download(blobName);
+    fs.unlinkSync(localFilePath);
+    await deleteDocumentFromAzure(blobName);
+    res.send();
```
resolved

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,477 @@
+    });
+  });
+} else {
+  mochaSuiteFn('tracing/cloud/azure/blob', function () {
```
resolved

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,462 @@
+/**
+ * endPoint is assigned with the default constant string for appending EndpointSuffix
+ * to connection string of an azure account
+ */
```
Please add https://learn.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string

athirasomanath replied:
resolved

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,462 @@
+          AZURE_CONTAINER_NAME: containerName,
+          AZURE_CONNECTION_STRING: connStr,
+          AZURE_STORAGE_ACCOUNT: storageAccount,
+          AZURE_ACCOUNT_KEY: accountKey
```
Please use either `AZURE` prefix or `AZ` prefix.

athirasomanath replied:
resolved

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,462 @@
+            path: '/upload'
+          });
+          await testUtils.retry(async () => {
+            const spans = await agentControls.getSpans();
```
Same here. See [comment](https://github.com/instana/nodejs/pull/967/files#r1444389715).

athirasomanath replied:
resolved

aryamohanan commented on packages/core/src/tracing/instrumentation/cloud/azure/blob.js:
```diff
@@ -0,0 +1,153 @@
+  instrumentBlobClient(blob.BlobClient);
+}
+
+// For the download and delete functionality applicable to block blobs, the BlockBlobClient extends BlobClient
```
Qs: Does the instrumentation only handle the following operations: upload, delete, download, and stageBlock?

athirasomanath replied:
yes. As of now it covers only the upload, delete and download ops. stageblock is one another method to upload

aryamohanan commented:
Good Job 🍾

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/app.mjs:
```diff
@@ -0,0 +1,326 @@
+/*
```
For now we want to stick to single mjs files 👌

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,328 @@
+    const downloadBlobResponse = await _blockBlobClient.downloadToFile(localFilePath);
+    return downloadBlobResponse;
+  } catch (e) {
+    log('Error occured while downloading:', e);
```
removed

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,328 @@
+const filePath = `${__dirname}/sample.pdf`;
+const localFilePath = `${__dirname}/out.pdf`;
+const data1 = fs.readFileSync(filePath);
+const accountKey = process.env.AZURE_ACCOUNT_KEY;
```
resolved

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,328 @@
+const request = require('request-promise-native');
+const filePath = `${__dirname}/sample.pdf`;
+const localFilePath = `${__dirname}/out.pdf`;
+const data1 = fs.readFileSync(filePath);
```
resolved

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,328 @@
+  blockBlobClient
+    .upload(pdfData)
+    .then(() => {
+      res.send('successss');
```
resolved

kirrg001 replied:
This was not a typo.

athirasomanath commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,462 @@
+          });
+          await testUtils.retry(async () => {
+            const spans = await agentControls.getSpans();
+            if (spans.length > 0) {
```
resolved

kirrg001 commented on packages/collector/test/tracing/cloud/azure/blob/test.js:
```diff
@@ -0,0 +1,462 @@
+ * which is the value for AZ_STORAGE_ACCOUNT_NAME. From the azure portal, navigate to this storage account
+ * and under the Access keys, Key can be found for AZ_STORAGE_ACCOUNT_KEY.
+ */
+if (!storageAccount || !accountKey) {
```
Yes good catch! Let's do it before merging.

kirrg001 commented on packages/core/src/tracing/instrumentation/cloud/azure/blob.js:
```diff
@@ -0,0 +1,153 @@
+
+function instrumentBlob(blob) {
+  instrumentBlockBlob(blob.BlockBlobClient);
+  instrumentBlobClient(blob.BlobClient);
```
issue: Why do we not instrument [this.client.sendOperationRequest](https://github.com/search?q=repo%3AAzure%2Fazure-sdk-for-js%20this.client.sendOperationRequest&type=code)?

athirasomanath replied:
Instrumenting BlockBlobClient and BlobClient give easy parsing for proper data required for span creation such as blobname, container name and operation. If using sendOperation, we need to parse it from the url and it doesn't seem neat. Also sendOperation is from @azure/core-http , and even though it is used in almost every operations in storage package, it doesn't act as common point of instrumentation when it comes to a major package like cosmosDB

aryamohanan commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,326 @@
+const app = express();
+const logPrefix = `Express / azure blob App (${process.pid}):\t`;
+const fs = require('fs');
+const request = require('request-promise-native');
```
please use `node-fetch` instead of `request-promise-native`.

athirasomanath replied:
changed

aryamohanan commented on packages/collector/test/tracing/cloud/azure/blob/app.mjs:
```diff
@@ -0,0 +1,328 @@
+const app = express();
+const logPrefix = `Express / azure blob App (${process.pid}):\t`;
+import fs from 'fs';
+import request from 'request-promise-native';
```
please use node-fetch instead of request-promise-native.

kirrg001 commented:
Please rebase against main and replace the usage of `ProcessControls`. The syntax has changed.
See example: https://github.com/instana/nodejs/commit/d1b38951f8d8111bdc2cce81aa1a361df4d1fc77#diff-92983afeee93cf42a78a3dd57e5a6e3e96662d66c7f7c6954548e0723322c601L34

athirasomanath commented:
> Hey!
> 
> I need to do another review iteration with fresh eyes. Please take a look at my first comments 👍
> 
> Here are some additional findings:
> 
> * ESM app is missing. Please add an ESM app.mjs and check if azure blob storage works with ESM - [see](https://github.com/instana/nodejs/blob/main/CONTRIBUTING.md#adding-tests).
> * I could not find the ticket reference in the PR nor in the commits.
> * _All of your files did not run through code formatting_ - please do.



> Hey!
> 
> I need to do another review iteration with fresh eyes. Please take a look at my first comments 👍
> 
> Here are some additional findings:
> 
> * ESM app is missing. Please add an ESM app.mjs and check if azure blob storage works with ESM - [see](https://github.com/instana/nodejs/blob/main/CONTRIBUTING.md#adding-tests).
> * I could not find the ticket reference in the PR nor in the commits.
> * _All of your files did not run through code formatting_ - please do.

Reply to this: Addressed the Issues

athirasomanath commented:
> Hey!
> 
> I need to do another review iteration with fresh eyes. Please take a look at my first comments 👍
> 
> Here are some additional findings:
> 
> * ESM app is missing. Please add an ESM app.mjs and check if azure blob storage works with ESM - [see](https://github.com/instana/nodejs/blob/main/CONTRIBUTING.md#adding-tests).
> * I could not find the ticket reference in the PR nor in the commits.
> * _All of your files did not run through code formatting_ - please do.

Reply to this: Addressed the Issues

kirrg001 commented:
Thank you @aryamohanan and me will do another review 👍

kirrg001 commented:
The plan is to review and merge this on Monday or better said: after the next release.

kirrg001 reviewed (approved):
LGTM!

Let's merge this after the full demo next week :)

kirrg001 commented:
Pushed a commit to run all tests using `[ci all-tests]`. Waiting for success result.

kirrg001 commented:
Passed 👍

aryamohanan commented on packages/collector/test/tracing/cloud/azure/blob/app.js:
```diff
@@ -0,0 +1,326 @@
+/*
+ * (c) Copyright IBM Corp. 2023
```
suggestion: 
> (c) Copyright IBM Corp. 2024

athirasomanath replied:
changed

aryamohanan reviewed (approved):
LGTM 🎉
